File: py/pipeline.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
MCP服务生成流水线
整合解析器、转换器和生成器，提供完整的API文档到MCP服务的转换流程
"""

import os
import time
from parsers import parse_document
from converters import convert_to_standard_format
from generators import generate_mcp_service
from utils.logger import get_logger

logger = get_logger()


class MCPPipeline:
    """
    MCP服务生成流水线类
    整合解析、转换和生成三个阶段
    """

    def __init__(self, output_dir="./O_MCP_SERVER_LIST"):
        """
        初始化MCP流水线

        Args:
            output_dir: 默认输出目录
        """
        self.output_dir = output_dir
        self.metrics = {
            "parse_time": 0,
            "convert_time": 0,
            "generate_time": 0,
            "total_time": 0,
            "endpoints_count": 0,
            "incomplete_endpoints": 0
        }

    def run(self, input_path, format="auto", service_name=None):
        """
        运行完整的MCP服务生成流水线

        Args:
            input_path: 输入文档路径，可以是URL或文件路径
            format: 输入文档格式，可选值为'swagger', 'openapi', 'markdown', 'auto'
            service_name: MCP服务名称，如果指定，将作为服务名并添加到输出目录后

        Returns:
            元组 (是否成功, 输出目录路径, 指标数据)
        """
        start_time = time.time()

        try:
            # 1. 解析阶段
            logger.info(f"📄 [1/3] 正在解析文档: {input_path}，格式: {format}")
            parse_start = time.time()
            api_data = self._parse(input_path, format)
            parse_end = time.time()
            self.metrics["parse_time"] = parse_end - parse_start

            if not api_data:
                logger.error("❌ 文档解析失败，流程终止")
                return False, None, self.metrics

            # 2. 转换阶段
            logger.info("🔄 [2/3] 正在转换为OpenAPI标准格式...")
            convert_start = time.time()
            standard_format = self._convert(api_data)
            convert_end = time.time()
            self.metrics["convert_time"] = convert_end - convert_start

            if not standard_format:
                logger.error("❌ 格式转换失败，流程终止")
                return False, None, self.metrics

            # 3. 处理服务名称
            output_dir = self._process_service_name(
                standard_format, service_name)

            # 4. 生成阶段
            logger.info(f"🚀 [3/3] 正在生成MCP服务到: {output_dir}")
            generate_start = time.time()
            success, metrics = self._generate(standard_format, output_dir)
            generate_end = time.time()
            self.metrics["generate_time"] = generate_end - generate_start

            # 更新指标
            self.metrics.update(metrics)
            self.metrics["total_time"] = time.time() - start_time

            if success:
                logger.info(f"✅ MCP服务生成完成，已写入到目录: {output_dir}")
                self._show_metrics()
                self._show_startup_guide(output_dir)
                return True, output_dir, self.metrics
            else:
                logger.error("❌ MCP服务生成失败")
                return False, output_dir, self.metrics

        except Exception as e:
            logger.error(f"❌ 流水线执行出错: {str(e)}")
            import traceback
            logger.debug(traceback.format_exc())

            self.metrics["total_time"] = time.time() - start_time
            return False, None, self.metrics

    def _parse(self, input_path, format):
        """解析阶段"""
        try:
            return parse_document(input_path, format)
        except Exception as e:
            logger.error(f"❌ 解析阶段出错: {str(e)}")
            return None

    def _convert(self, api_data):
        """转换阶段"""
        try:
            return convert_to_standard_format(api_data)
        except Exception as e:
            logger.error(f"❌ 转换阶段出错: {str(e)}")
            return None

    def _process_service_name(self, standard_format, service_name):
        """处理服务名称"""
        if service_name:
            logger.info(f"🏷️ 使用指定的服务名称: {service_name}")
            # 设置服务名称
            if 'title' not in standard_format:
                standard_format['title'] = service_name
            else:
                standard_format['original_title'] = standard_format.get(
                    'title')
                standard_format['title'] = service_name

            # 将服务名添加到输出目录
            output_dir = os.path.join(self.output_dir, service_name)
        else:
            # 尝试从标准格式中获取服务名
            if 'title' in standard_format and standard_format['title']:
                service_title = standard_format['title']
                # 清理标题，使其适合作为目录名
                service_title = service_title.replace(' ', '_').lower()
                output_dir = os.path.join(self.output_dir, service_title)
                logger.info(f"🏷️ 使用API标题作为服务名称: {service_title}")
            else:
                output_dir = self.output_dir

        # windows下替换路径
        if os.name == 'nt':
            output_dir = output_dir.replace('\\', '/')
        return output_dir

    def _generate(self, standard_format, output_dir):
        """生成阶段"""
        try:
            success = generate_mcp_service(standard_format, output_dir)

            # 获取生成的信息
            metrics = {
                "endpoints_count": 0,
                "incomplete_endpoints": 0
            }

            # 如果是OpenAPI格式，从paths统计端点数量
            if 'paths' in standard_format:
                endpoint_count = 0
                for path_item in standard_format['paths'].values():
                    for method in ['get', 'post', 'put', 'delete', 'patch', 'options', 'head']:
                        if method in path_item:
                            endpoint_count += 1
                metrics["endpoints_count"] = endpoint_count
            elif 'endpoints' in standard_format:
                metrics["endpoints_count"] = len(standard_format['endpoints'])

            return success, metrics
        except Exception as e:
            logger.error(f"❌ 生成阶段出错: {str(e)}")
            return False, {}

    def _show_metrics(self):
        """显示性能指标"""
        # 合并指标数据到一个info log
        logger.info("\n📊 流水线执行指标:"
                    + f"\n总执行时间: {self.metrics['total_time']:.2f}秒"
                    + f"\n总执行时间: {self.metrics['total_time']:.2f}秒"
                    + f"\n解析阶段: {self.metrics['parse_time']:.2f}秒"
                    + f"\n转换阶段: {self.metrics['convert_time']:.2f}秒"
                    + f"\n生成阶段: {self.metrics['generate_time']:.2f}秒"
                    + f"\n端点总数: {self.metrics['endpoints_count']}个")

    def _show_startup_guide(self, output_dir):
        """显示启动指南"""
        mcp_server_path = os.path.join(output_dir, 'mcp_server.py')
        # windows下替换路径
        if os.name == 'nt':
            mcp_server_path = mcp_server_path.replace('\\', '/')
        logger.info(
            f"""\n🚀 启动MCP服务:\n
1. 直接运行生成的Python文件
> python {mcp_server_path}

2. 使用MCP CLI工具运行测试
> mcp dev {mcp_server_path}

3. 使用环境变量自定义API基础URL
> API_BASE_URL=https://your-api-base-url python {mcp_server_path}"""
        )


def run_pipeline(input_path, format="auto", output_dir="./O_MCP_SERVER_LIST", service_name=None):
    """
    运行MCP服务生成流水线的便捷函数

    Args:
        input_path: 输入文档路径，可以是URL或文件路径
        format: 输入文档格式，可选值为'swagger', 'openapi', 'markdown', 'auto'
        output_dir: 输出目录
        service_name: MCP服务名称，如果指定，将作为服务名并添加到输出目录后

    Returns:
        元组 (是否成功, 输出目录路径, 指标数据)
    """
    pipeline = MCPPipeline(output_dir)
    return pipeline.run(input_path, format, service_name)

File: ts/src/pipeline.ts
/**
 * MCP service generation pipeline
 * Integrates parsers, converters and generators to provide a complete
 * API documentation to MCP service conversion workflow
 */

import * as path from 'path';
import { convertToStandardFormat } from './converters';
import { generateMcpService } from './generators';
import { parseDocument } from './parsers';
import logger from './utils/logger';

/**
 * Metrics about the pipeline execution
 */
interface PipelineMetrics {
  parseTime: number;
  convertTime: number;
  generateTime: number;
  totalTime: number;
  endpointsCount: number;
  incompleteEndpoints: number;
}

/**
 * MCP service generation pipeline class
 */
export class MCPPipeline {
  private outputDir: string;
  private metrics: PipelineMetrics;

  /**
   * Initialize the MCP pipeline
   * @param outputDir Default output directory
   */
  constructor(outputDir: string = "./O_MCP_SERVER_LIST") {
    this.outputDir = outputDir;
    this.metrics = {
      parseTime: 0,
      convertTime: 0,
      generateTime: 0,
      totalTime: 0,
      endpointsCount: 0,
      incompleteEndpoints: 0
    };
  }

  /**
   * Run the complete MCP service generation pipeline
   * @param inputPath Input document path, can be a URL or file path
   * @param format Input document format, can be 'swagger', 'openapi', 'markdown', 'auto'
   * @param serviceName MCP service name, if specified, will be used as service name and added to output directory
   * @returns Tuple of (success, output directory path, metrics data)
   */
  async run(
    inputPath: string, 
    format: string = "auto", 
    serviceName?: string
  ): Promise<[boolean, string | null, PipelineMetrics]> {
    const startTime = Date.now();

    try {
      // 1. Parse phase
      logger.info(`📄 [1/3] Parsing document: ${inputPath}, format: ${format}`);
      const parseStart = Date.now();
      const apiData = await this._parse(inputPath, format);
      const parseEnd = Date.now();
      this.metrics.parseTime = (parseEnd - parseStart) / 1000;

      if (!apiData) {
        logger.error("❌ Document parsing failed, process terminated");
        return [false, null, this.metrics];
      }

      // 2. Convert phase
      logger.info("🔄 [2/3] Converting to OpenAPI standard format...");
      const convertStart = Date.now();
      const standardFormat = await this._convert(apiData);
      const convertEnd = Date.now();
      this.metrics.convertTime = (convertEnd - convertStart) / 1000;

      if (!standardFormat) {
        logger.error("❌ Format conversion failed, process terminated");
        return [false, null, this.metrics];
      }

      // 3. Process service name
      const outputDir = this._processServiceName(standardFormat, serviceName);

      // 4. Generate phase
      logger.info(`🚀 [3/3] Generating MCP service to: ${outputDir}`);
      const generateStart = Date.now();
      const [success, generatorMetrics] = await this._generate(standardFormat, outputDir);
      const generateEnd = Date.now();
      this.metrics.generateTime = (generateEnd - generateStart) / 1000;

      // Update metrics
      this.metrics = { ...this.metrics, ...generatorMetrics };
      this.metrics.totalTime = (Date.now() - startTime) / 1000;

      if (success) {
        logger.info(`✅ MCP service generation complete, written to directory: ${outputDir}`);
        this._showMetrics();
        this._showStartupGuide(outputDir);
        return [true, outputDir, this.metrics];
      } else {
        logger.error("❌ MCP service generation failed");
        return [false, outputDir, this.metrics];
      }
    } catch (e) {
      const error = e as Error;
      logger.error(`❌ Pipeline execution error: ${error.message}`);
      logger.debug(error.stack || "No stack trace available");

      this.metrics.totalTime = (Date.now() - startTime) / 1000;
      return [false, null, this.metrics];
    }
  }

  /**
   * Parse phase
   * @param inputPath Input document path
   * @param format Document format
   * @returns Parsed API data
   */
  private async _parse(inputPath: string, format: string): Promise<any> {
    try {
      return await parseDocument(inputPath, format);
    } catch (e) {
      const error = e as Error;
      logger.error(`❌ Parse phase error: ${error.message}`);
      return null;
    }
  }

  /**
   * Convert phase
   * @param apiData API data to convert
   * @returns Standard format data
   */
  private async _convert(apiData: any): Promise<any> {
    try {
      return await convertToStandardFormat(apiData);
    } catch (e) {
      const error = e as Error;
      logger.error(`❌ Convert phase error: ${error.message}`);
      return null;
    }
  }

  /**
   * Process service name
   * @param standardFormat Standard format data
   * @param serviceName Service name
   * @returns Output directory path
   */
  private _processServiceName(standardFormat: any, serviceName?: string): string {
    let outputDir: string;

    if (serviceName) {
      logger.info(`🏷️ Using specified service name: ${serviceName}`);
      
      // Set service name
      if (!standardFormat.title) {
        standardFormat.title = serviceName;
      } else {
        standardFormat.originalTitle = standardFormat.title;
        standardFormat.title = serviceName;
      }

      // Add service name to output directory
      outputDir = path.join(this.outputDir, serviceName);
    } else {
      // Try to get service name from standard format
      if (standardFormat.title && typeof standardFormat.title === 'string') {
        const serviceTitle = standardFormat.title.replace(/\s+/g, '_').toLowerCase();
        outputDir = path.join(this.outputDir, serviceTitle);
        logger.info(`🏷️ Using API title as service name: ${serviceTitle}`);
      } else {
        outputDir = this.outputDir;
      }
    }

    // Handle Windows paths
    if (process.platform === 'win32') {
      outputDir = outputDir.replace(/\\/g, '/');
    }
    
    return outputDir;
  }

  /**
   * Generate phase
   * @param standardFormat Standard format data
   * @param outputDir Output directory
   * @returns Tuple of (success, metrics)
   */
  private async _generate(standardFormat: any, outputDir: string): Promise<[boolean, Partial<PipelineMetrics>]> {
    try {
      const success = await generateMcpService(standardFormat, outputDir);

      // Get generated information
      const metrics: Partial<PipelineMetrics> = {
        endpointsCount: 0,
        incompleteEndpoints: 0
      };

      // If OpenAPI format, count endpoints from paths
      if (standardFormat.paths) {
        let endpointCount = 0;
        for (const pathItem of Object.values(standardFormat.paths)) {
          const pathObj = pathItem as any;
          for (const method of ['get', 'post', 'put', 'delete', 'patch', 'options', 'head']) {
            if (method in pathObj) {
              endpointCount += 1;
            }
          }
        }
        metrics.endpointsCount = endpointCount;
      } else if (standardFormat.endpoints) {
        metrics.endpointsCount = standardFormat.endpoints.length;
      }

      return [success, metrics];
    } catch (e) {
      const error = e as Error;
      logger.error(`❌ Generate phase error: ${error.message}`);
      return [false, {}];
    }
  }

  /**
   * Show performance metrics
   */
  private _showMetrics(): void {
    // Combine metrics data into a single info log
    logger.info(
      "\n📊 Pipeline execution metrics:" +
      `\nTotal execution time: ${this.metrics.totalTime.toFixed(2)} seconds` +
      `\nParse phase: ${this.metrics.parseTime.toFixed(2)} seconds` +
      `\nConvert phase: ${this.metrics.convertTime.toFixed(2)} seconds` +
      `\nGenerate phase: ${this.metrics.generateTime.toFixed(2)} seconds` +
      `\nTotal endpoints: ${this.metrics.endpointsCount}`
    );
  }

  /**
   * Show startup guide
   * @param outputDir Output directory
   */
  private _showStartupGuide(outputDir: string): void {
    const mcpServerPath = path.join(outputDir, 'mcp_server.js');
    // Handle Windows paths
    const formattedPath = process.platform === 'win32' ? 
      mcpServerPath.replace(/\\/g, '/') : 
      mcpServerPath;

    logger.info(
      `\n🚀 Start MCP service:\n
1. Run the generated JavaScript file
> node ${formattedPath}

2. Use MCP CLI tool to run tests
> mcp dev ${formattedPath}

3. Customize API base URL with environment variables
> API_BASE_URL=https://your-api-base-url node ${formattedPath}`
    );
  }
}

/**
 * Convenience function to run the MCP service generation pipeline
 * @param inputPath Input document path, can be a URL or file path
 * @param format Input document format, can be 'swagger', 'openapi', 'markdown', 'auto'
 * @param outputDir Output directory
 * @param serviceName MCP service name, if specified, will be used as service name and added to output directory
 * @returns Tuple of (success, output directory path, metrics data)
 */
export async function runPipeline(
  inputPath: string, 
  format: string = "auto", 
  outputDir: string = "./O_MCP_SERVER_LIST", 
  serviceName?: string
): Promise<[boolean, string | null, PipelineMetrics]> {
  const pipeline = new MCPPipeline(outputDir);
  return await pipeline.run(inputPath, format, serviceName);
} 